# HW3 Part 2 MGT 6203 

## Instructions:


**For Homework 3 Part 2, please use this Jupyter notebook in Vocareum to submit your solutions.** Vocareum is an educational cloud platform for programming in several languages; it is based on the Jupyter notebook environment. This platform allows us to move homework assignments to the cloud. The advantages are that all of you will be working in the same coding environment AND peer reviewers will be able to run your R code easily. This way we eliminate some issues we might encounter when working on an individual/local basis, such as library installations and Rstudio OS requirements; Jupyter notebooks work on mobile platforms and tablets.

With Jupyter notebooks, you will learn a new way of presenting data analysis reports that is neat and flexible, where formatted English text and R code can easily coexist on the same page. Notebooks can be also collaborative when needed. Think of Jupyter notebooks as interactive program-based Google docs or MS-Office 360 docs; these are gradually replacing local files on our computers. **For now, we are asking each of you to do your own homework.**

**Many of you are new to the Jupyter notebooks and Vocareum platforms. We will provide TA help in Piazza with specific code if you have questions. ** 

Here we list some important things to get you started. Please read carefully.



<ol>
<li>Even though we are moving from your local envrionment to the cloud, **our expectations from your homework will remain the same**. Same goes for the rubrics.</li>

<li>Vocareum has its own cloud based file system, the data files you will be using for the assignments will be stored in the cloud with path **"../resource/asnlib/publicdata/FILENAME.csv"**. You will be able to import them with the same method as you do in RStudio, simply substitute the path name to the one specified in the instructions. You won't be able to modify these data files.</li>

<li>You will be able to find the data files on Canvas/EdX if you would like to explore them offline.</li>

<li>**For coding questions, you will be graded on the R code as well as the output** in your submission.</li>

<li>**For interpretations or short response questions, please type the answers in the notebook's markdown cells**. To change a code cell to a markdown cell, click on the cell, and in the dropdown menu above, switch the type of the cell block from "code" to "markdown". **Adding print statements to code cells for short response/interpretation questions is also fine, as long as we can clearly see the output of your response**.</li>

<li>Optional: if you would like to learn more about how to format your markdown cells, visit the following site: https://www.earthdatascience.org/courses/intro-to-earth-data-science/file-formats/use-text-files/format-text-with-markdown-jupyter-notebook/ </li>

<li> Jupyter notebook also supports LaTeX. </li>
<li> If you'd like to download our datasets from Vocareum for use offline, you're welcome to follow the steps below:
    <ol>
    <li> Run the code cell that reads the data into Vocareum and assigns it to a variable. </li>
    <li> Write `write.csv(data_read_into_vocareum_variable, "file_name.csv")` into a code cell, and run it. </li>
    <li> Save your work by clicking on the floppy disk icon in the top-left. </li>
    <li> After saving your work, in Vocareum, navigate to `File` -> `Open`. </li>
    <li> You'll then see the file name, `file_name.csv`, produced from the `write.csv` function. </li>
    <li> Click on it to download. </li>
    </ol>
</li>


<li>**Feel free to delete or add as many additional cells as you need**. But please try to keep your notebook clean and keep your solution directly under its respective question to avoid confusion.</li>

<li>You may delete the `#SOLUTION BEGINS/ENDS HERE` comments from the cell blocks, they are just pointers that indicate where to put your solutions.</li>

<li>**When you have finished the assignment, remember to rerun your notebook to check if it runs correctly.** You can do so by going to **`Kernel`-> `Restart & Run All`**. You may lose points if your solutions do not run successfully.</li>

<li>**Click the "`Submit`" button on the top right corner to turn in your assignment**. Your assignment will enter the next phase for peer review.</li>

<li>**You are allowed a total of <u>2 submissions</u> for this assignment.** So make sure that you submit your responses carefully. You will be able to come back and resubmit your assignment as long as it is before the start of the peer review period.</li>

<li>**Please remember to finish the peer reviews after having submitted your assignment.** You are responsible for grading the work of three of your peers thoroughly, and in adherence to the rubrics. And you will be held accountable for peer grading. **There will be a <u>30% penalty</u> to your grade if you fail to complete one or more peer reviews properly.**</li>

<li>Feel free to address your questions, concerns, and provide any feedback on Piazza. We will continuously try to improve going forward.</li>

<li>Good Luck!</li>

</ol>



# About Package Installation:

Most of the packages (if not all) that you will need to complete this assignment are already installed in this environment. An easy way to check this is to run the command: `library(PackageName)`. If this command runs successfully then the package was already installed and has been successfully attached to the code. If the command gave an error saying the Package was not found then follow the steps below to successfully install the package and attach it to the code: 

- Use `installed.packages()` command to return a table of the packages that are preinstalled in the environment.

- To attach a preinstalled library in Vocareum, simply use `library(PackageName)`

- **To install a package that does not come with the provided environment, please use the following steps:**
    
<ol>
<li>Install your desired package using the following code:</li>
     <ul>
         <li>`install.packages("PackageName", lib="../work/")`</li>
     </ul>
<li> Attach the library you just installed using the following code: </li>
    <ul>
        <li>`library(PackageName, lib.loc = "../work/")`</li>
        <li> **NOTE:** Make sure the file location is the same as the above code snippets `("../work/")` </li>
    </ul>
</ol>

## Instructions for using the `safe_read_csv()` function

- **Purpose:**  The `safe_read_csv()` function is designed to verify the existence of datasets for specific questions.
   
- **Execution:**

    + <span style="color:blue"> **Run the function to attempt loading the dataset into your Vocareum notebook.This function only needs to be run once each time you open the notebook** </span>:
      - If the dataset loads successfully, you will see ** a "...successfully read" message.** Please proceed with the rest of the questions.
      -  If the dataset is missing, an error message stating **"…not found…"**  will appear.Please report this issue by posting on Piazza so we can resolve it.
      
    

In [3]:
# The safe_read_csv function is used for checking datasets.
safe_read_csv <- function(file_path, stringsAsFactors = TRUE) {
    result <- tryCatch({
        data <- read.csv(file_path, stringsAsFactors = stringsAsFactors)
        mes <- paste(file_path, " successfully read.")
        print(mes)
        return(data)
      }, warning = function(w) {
    # Warning Message
        message("Warning: ", conditionMessage(w))
        message("Using alternative file path.")
        # Use regular expression to get file name after last slash
        file_name <- sub("^.*/", "", file_path) 
        # See if file in first layer of working directory.
        # If not, student can easily upload it to Vocareum directory.
        result <- read.csv(file_name, stringsAsFactors = stringsAsFactors)
        print("Alternate path successful")
        return(result)
    }, error = function(e) {
    # If neither method was successful
        message("Error: ", conditionMessage(e))
        print("Please upload data to Vocareum")
      }, finally = {
        ## This message appears regardless of whether an error appears
        final_mes <- paste("If ", file_path, " not found, please notify us in Piazza. Thank you.")
        print(final_mes)
    }) 
    return(result)
}

## Instructions for Question 1 (16 Points)

For Question 1, the goal is to use the data to construct a three-factor model. We'll do this by looking at how returns for `UPS` and `KO`, after accounting for the risk-free rate, relate to three factors—`Mkt_RF`, `SMB`, and `HML`—during our study period. Specifically, you  will employ regression analysis to conduct factor investing for UPS and KO stocks.

### Please use the dataset '`UPS_KO.csv`' to answer the following questions. 

The data set contains the following columns:

- **`Date`**: This column represents date from 09/2014 to 08/2019.
- **`Mkt_RF`**: This column represents market premium (i.e., Market return – risk_free rate).
- **`SMB`**: This column represents the value of the size factor.
- **`HML`**: This column represents the value of the value factor.
- **`RF`**: This column represents risk free rate.
- **`UPS`**: This column represents the daily return of UPS.
- **`KO`**: This column represents the daily return of KO.

Sample Period : 04/2015 to 11/2018 (Inclusive)




In [4]:
# Load the data and necessary library
suppressPackageStartupMessages(library(dplyr))
library(dplyr)

# Read the dataset
# data_Q1 <- read.csv("../resource/asnlib/publicdata/UPS_KO.csv")
data_Q1 <- safe_read_csv("../resource/asnlib/publicdata/UPS_KO.csv")

# Format the data
data_Q1 <- data_Q1 %>% mutate(Date = as.Date(paste0(Date, "01"), format = "%Y%m%d"))

# Define the start and end dates for filtering
start_date <- as.Date("2015-04-01")
end_date <- as.Date("2018-11-30")

# Filter the dataframe based on the date range
filtered_df <- data_Q1 %>%
  filter(Date >= start_date & Date <= end_date) %>% 
  arrange(Date)

filtered_df$UPS_RF <- filtered_df$UPS - filtered_df$RF
filtered_df$KO_RF <- filtered_df$KO - filtered_df$RF


tail(filtered_df)

[1] "../resource/asnlib/publicdata/UPS_KO.csv  successfully read."
[1] "If  ../resource/asnlib/publicdata/UPS_KO.csv  not found, please notify us in Piazza. Thank you."


Date       Mkt_RF  SMB     HML     RF     UPS         KO         
39 2018-06-01  0.0048  0.0117 -0.0239 0.0014 -0.07797406  0.02000000
40 2018-07-01  0.0319 -0.0215  0.0039 0.0016  0.12858888  0.07262409
41 2018-08-01  0.0344  0.0123 -0.0412 0.0016  0.02493950 -0.04417764
42 2018-09-01  0.0006 -0.0237 -0.0130 0.0015 -0.04269296  0.03634731
43 2018-10-01 -0.0768 -0.0476  0.0344 0.0019 -0.08745183  0.04540521
44 2018-11-01  0.0169 -0.0079  0.0025 0.0018  0.08212880  0.05263169
   UPS_RF      KO_RF      
39 -0.07937405  0.01860000
40  0.12698888  0.07102409
41  0.02333950 -0.04577764
42 -0.04419296  0.03484731
43 -0.08935183  0.04350521
44  0.08032880  0.05083169

### A. Using the dataset named `filtered_df` provided above, fit a factor regression model for UPS over the sample period using `Mkt_RF`, `SMB`, and `HML` (6 points). 


- Report each of the coefficients and interpret each variable (including intercept) according to their financial interpretation AND taking statistical significance into account at alpha=0.1.

In [5]:
# SOLUTION BEGINS HERE
UPS_factor <- lm(UPS ~ Mkt_RF + SMB + HML, data = filtered_df)
summary(UPS_factor)

coefficients <- round(summary(UPS_factor)$coefficients[, "Estimate"], 4)
coefficients
# SOLUTION ENDS HERE


Call:
lm(formula = UPS ~ Mkt_RF + SMB + HML, data = filtered_df)

Residuals:
      Min        1Q    Median        3Q       Max 
-0.135574 -0.024485 -0.000379  0.026267  0.093810 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -7.249e-05  7.037e-03  -0.010   0.9918    
Mkt_RF       1.065e+00  2.311e-01   4.607 4.11e-05 ***
SMB          4.331e-02  2.776e-01   0.156   0.8768    
HML          4.669e-01  2.754e-01   1.695   0.0978 .  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.04492 on 40 degrees of freedom
Multiple R-squared:  0.3759,	Adjusted R-squared:  0.3291 
F-statistic: 8.032 on 3 and 40 DF,  p-value: 0.0002626


(Intercept)      Mkt_RF         SMB         HML 
    -0.0001      1.0646      0.0433      0.4669

Coefficient Interpretations:
(Intercept) - The fund manager underpeformed during this time period, but the coefficient is not significant at alpha = 0.10.
Mkt_RF - The fund has a market beta of -7.249e-05 or -0.00007249 and is very significant to UPS daily returns.
SMB - The fund favored small stocks but not statistically relevant to performance.
HML - The fund favored higher value stocks, which is signifiacnt at the given alpha level.

### B. Using the dataset named `filtered_df` provided above, fit a factor regression model for KO over the sample period using `Mkt_RF`, `SMB`, and `HML` (6 points). 
- Report each of the coefficients (including intercept) and interpret each according to their financial interpretation AND taking statistical significance into account at alpha=0.1

In [6]:
# SOLUTION BEGINS HERE
KO_factor <- lm(KO ~ Mkt_RF + SMB + HML, data = filtered_df)
summary(KO_factor)

coefficients <- round(summary(KO_factor)$coefficients[, "Estimate"], 4)
coefficients
# SOLUTION ENDS HERE


Call:
lm(formula = KO ~ Mkt_RF + SMB + HML, data = filtered_df)

Residuals:
      Min        1Q    Median        3Q       Max 
-0.076463 -0.018782  0.003381  0.018957  0.046478 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept)  0.003884   0.004347   0.894  0.37692    
Mkt_RF       0.460956   0.142765   3.229  0.00249 ** 
SMB         -0.799243   0.171479  -4.661 3.47e-05 ***
HML          0.015482   0.170118   0.091  0.92794    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.02775 on 40 degrees of freedom
Multiple R-squared:  0.4052,	Adjusted R-squared:  0.3606 
F-statistic: 9.082 on 3 and 40 DF,  p-value: 0.000104


(Intercept)      Mkt_RF         SMB         HML 
     0.0039      0.4610     -0.7992      0.0155

Coefficient Interpretations:
(Intercept) - The fund manager overpeformed during this time period but not statistically relevant.
Mkt_RF - The fund has a market beta of 0.460956 and is significant to KO daily returns.
SMB - The fund favored larger stocks and this is very significant to KO performance
HML - The fund favored higher value stocks, but this is not signifiacnt at the given alpha level.

### C. Based on Warren Buffett's value investing strategy which focuses on financial health and growth potential, which company is he more likely to invest in according to these factor model outputs? (4 points)

#### SOLUTION BEGINS HERE (Double click to edit)

Warren Buffett's investing strategy involves high value, large cap stocks. This would correlated to positive and significant HML values as well as negative and significant SMB values. That said, Buffett is more likely to invest in KO, as it also favored large, high value stocks during this time period. Further, KO also sees that these factors are much more statistically significant to KO's daily returns. 

#### SOLUTION ENDS HERE

## Instructions for Question 2 (32 Points)

For Question 2, you will use logistic regression to segment customers into two classes.  This question is adapted from a Kaggle contest to evaluate current customers for an auto dealership that is opening a new location.  The dealership wishes to reliably segment customers in order to streamline marketing efforts when their new location opens.  
<br />
The data for these questions is contained in the file, **`auto_customers_two_segments.csv`**, which includes the following fields:

- **`ID`** : Unique ID created by dealership (Removed from `data_Q2`)
- **`Gender`** : Gender of the customer
- **`Ever_Married`** : Marital status of the customer
- **`Age`** : Age of the customer
- **`Graduated`** : Is the customer a college graduate?
- **`Profession`** : Profession of the customer
- **`Work_Experience`** : Work experience in years
- **`Spending_Score`** : Spending score of the customer.
- **`Family_Size`** : Number of family members (including customer)
- **`Segmentation`** : The marketing segment the customer is in.



In [7]:
# Load necessary library and data
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(knitr))
library(caret)

#load data from file
data_Q2 = safe_read_csv('../resource/asnlib/publicdata/auto_customers_two_segments.csv')
head(data_Q2 )

Loading required package: lattice


Attaching package: ‘caret’


The following object is masked from ‘package:purrr’:

    lift




[1] "../resource/asnlib/publicdata/auto_customers_two_segments.csv  successfully read."
[1] "If  ../resource/asnlib/publicdata/auto_customers_two_segments.csv  not found, please notify us in Piazza. Thank you."


Gender Ever_Married Age Graduated Profession    Work_Experience
1 Male   No           22  No        Healthcare     1             
2 Female Yes          38  Yes       Engineer      NA             
3 Female Yes          40  Yes       Entertainment NA             
4 Female No           33  Yes       Healthcare     1             
5 Female Yes          61  Yes       Engineer       0             
6 Female No           26  Yes       Engineer       1             
  Spending_Score Family_Size Segmentation
1 Low            4           B           
2 Average        3           A           
3 High           6           A           
4 Low            3           B           
5 Low            3           B           
6 Low            3           A

### A. Follow the steps below to clean the data set and answer the question.   (4 Points)


1. Remove all rows with `n/a` values. 
2. Remove all rows with data points that are blank (**these are "" values** that have a string length of 0).  You can use `dplyr` or other methods to complete this step. (**HINT:** the cleaned sample size is 3400)
3. What are the customer segment labels (**`Segmentation`** column) and how many customers are in each segment? 



In [8]:
# SOLUTION BEGINS HERE
data_Q2 <- data_Q2  %>% 
    mutate(across(where(is.factor), as.character)) %>%
    filter(if_all(everything(), ~ . != "NA" & . != ""))
nrow(data_Q2)

labels <- unique(data_Q2$Segmentation)
labels

a_customers <- filter(data_Q2, Segmentation == "A")
b_customers <- filter(data_Q2, Segmentation == "B")

paste("# of A customers:", nrow(a_customers))
paste("# of B customers:", nrow(b_customers))
paste("Total # of customers:", sum(c(nrow(a_customers), nrow(b_customers))))

#data_Q2 <- data_Q2  %>%
#    mutate(across(where(is.character), as.factor))
#str(data_Q2)
# SOLUTION ENDS HERE

[1] 3400

[1] "B" "A"

[1] "# of A customers: 1628"

[1] "# of B customers: 1772"

[1] "Total # of customers: 3400"

### B. Create a logistic regression model using `Segmentation` as the response variable and all other variables as the predictors. (4 Points)
- Segment A represents the most desirable customers
- Segment B represents customers who have made purchases, but are not as likely to have a high or average spending score.  
- **HINT: You need to specify Segment B as a reference level by setting Segment B to 0 and Segment A to 1.**

In [9]:
# SOLUTION BEGINS HERE
data_Q2 <- data_Q2  %>% 
    mutate(across(where(is.character), as.factor))  %>% 
    mutate(Segmentation = ifelse(Segmentation == 'B', 0, 1))

segmenting_model <- glm(Segmentation ~ Gender + Ever_Married + Age + Graduated + Profession + Work_Experience + Spending_Score + Family_Size, family = 'binomial', data = data_Q2)
summary(segmenting_model)
# SOLUTION ENDS HERE


Call:
glm(formula = Segmentation ~ Gender + Ever_Married + Age + Graduated + 
    Profession + Work_Experience + Spending_Score + Family_Size, 
    family = "binomial", data = data_Q2)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.3515  -0.8024  -0.2977   0.8327   2.6163  

Coefficients:
                         Estimate Std. Error z value Pr(>|z|)    
(Intercept)              1.160662   0.278123   4.173 3.00e-05 ***
GenderMale              -0.371184   0.094701  -3.920 8.87e-05 ***
Ever_MarriedYes          0.428553   0.118130   3.628 0.000286 ***
Age                      0.021640   0.004137   5.231 1.68e-07 ***
GraduatedYes             0.617568   0.089949   6.866 6.61e-12 ***
ProfessionDoctor        -1.125080   0.161168  -6.981 2.94e-12 ***
ProfessionEngineer      -0.847733   0.166809  -5.082 3.73e-07 ***
ProfessionEntertainment -0.669584   0.147764  -4.531 5.86e-06 ***
ProfessionExecutive     -1.450695   0.209493  -6.925 4.37e-12 ***
ProfessionHealthcare    -

### C. Use the cleaned data set  for both training and testing data and make predictions of the probability for each customer. (4 Points)

- For simplicity, we are using the same dataset for both training and testing.
- Display the first 5 values. **Note: round your results to three decimal places, e.g., XX.xxx.**
- **HINT**: You can use the **`predict()`** function on the model with the dataset to generate a vector of probability predictions.


In [10]:
# SOLUTION BEGINS HERE
prob_predictions <- predict(segmenting_model, data_Q2, type = "response")
data_Q2 <- data_Q2  %>% 
    mutate(predicted_probs = prob_predictions)

head(data_Q2 %>% select(predicted_probs) %>% mutate(predicted_probs = round(predicted_probs, 3)), 5)
# SOLUTION ENDS HERE

predicted_probs
1 0.052          
2 0.174          
3 0.821          
4 0.579          
5 0.046

### D. Predict the accuracy of the model for each of the following cutoff probability values. (4 Points)
- Classify each predicted value as 1 (representing Segment A) or 0 (representing Segment B), using the following cutoff probability values: 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8.
- Calculate accuracy: $$ \text{Accuracy} = \frac{(TP + TN)}{(TP + TN + FP + FN)}$$
- **Round your results to three decimal places, e.g., XX.xxx.**
-  **Note: You will need to calculate the accuracy seven times, once for each cutoff value.**


In [11]:
# SOLUTION BEGINS HERE
cutoff_vals <- c(0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8)
accuracies <- numeric(length(cutoff_vals))
sensitivities <- numeric(length(cutoff_vals))
specificities <- numeric(length(cutoff_vals))

for (i in seq_along(cutoff_vals)) {
    cutoff <- cutoff_vals[i]
    data_Q2 <- data_Q2  %>% 
                    mutate(predicted_segment = ifelse(predicted_probs >= cutoff, 1, 0))  %>% 
                    mutate(Segmentation = ifelse(Segmentation == 0, "B", "A"))

    TP <- sum(data_Q2$predicted_segment == 1 & data_Q2$Segmentation == "A")
    TN <- sum(data_Q2$predicted_segment == 0 & data_Q2$Segmentation == "B")
    FP <- sum(data_Q2$predicted_segment == 1 & data_Q2$Segmentation == "B")
    FN <- sum(data_Q2$predicted_segment == 0 & data_Q2$Segmentation == "A")

    accuracy <- (TP + TN) / (TP + TN + FP + FN)
    sensitivity <- (TP) / (TP + FN)
    specificity <-(TN) / (TN + FP)
    
    accuracies[i] <- accuracy
    sensitivities[i] <- sensitivity
    specificities[i] <- specificity
}

round(accuracies, 3)
# SOLUTION ENDS HERE

[1] 0.692 0.681 0.608 0.523 0.427 0.297 0.146

### E. Which cutoff probability value ryields the highest accuracy for the model? What is the accuracy as a percentage? (4 Points)
- Round percentages to two decimal places, for example, 80.51%


In [12]:
# SOLUTION BEGINS HERE

paste("The first model with a cutoff probability value of 0.2 yields the highestest accuracy. As a percentage, it is an accuracy of", round(accuracies[1]*100, 2),"%.")

# SOLUTION ENDS HERE

[1] "The first model with a cutoff probability value of 0.2 yields the highestest accuracy. As a percentage, it is an accuracy of 69.21 %."

### F. Now we have developed a model and found the model that produces the highest accuracy. Create a confusion matrix using predictions from the optimal cutoff probability value above. (4 Points)
- **HINT**: You may use the caret package function, **`confusionMatrix()`** to automate this process. 

In [13]:
# SOLUTION BEGINS HERE
optimal_cutoff <- 0.2
data_Q2 <- data_Q2  %>% mutate(Segmentation = factor(Segmentation, levels = c("A", "B"))) %>% 
                        mutate(predicted_class = factor(ifelse(predicted_probs >= optimal_cutoff, "A", "B"), levels = c("A", "B")))

confusion <- matrix(c(TP, TN, FP, FN), nrow = 2, ncol = 2)
confusion
# SOLUTION ENDS HERE

[,1] [,2]
[1,] 495     0
[2,]   0  2905

### G. Report the accuracy, sensitivity and specificity of the chosen model as percentages. (4 Points)
- Round percentages to two decimal places, for example, 80.51%


In [14]:
# SOLUTION BEGINS HERE
round(c(accuracies[1]*100, sensitivities[1]*100, specificities[1]*100), 2)
# SOLUTION ENDS HERE

[1] 69.21 95.21 45.32

### H. Evaluate your model. (4 Points) 
After you shared your results with the auto dealership's leadership team, they expressed concerns about the high number of customers from Segment A being misclassified as Segment B.
 
Please address the following specific questions:
- Why is misclassifying Segment A customers as Segment B problematic? What implications does this have?
- How would you adjust your model to more accurately classify Segment A customers?
- What are the potential downsides of making these adjustments to your model?

#### SOLUTION BEGINS HERE (Double click to edit)

Misclassifying Segment A customers into Segment B is a a type II error, or a false negative. The implications of such a large error can cost the auto dealership lots of money. If the model shows that most people are NOT likely to purchase from the company, marketing may shift away from what the company was already doing, and into something completely different, something they think will bring more customers, but because of the many false negatives, company executives may not see the consequences of making such changes until after financial damage is significant. One thing I may do to adjust this model is to add an interaction term as an independent variable, for example, Age_Spending_Score may be defined as Age * Spending_Score. This adjustment has one major potential downside, which is that interaction terms can sometimes encourage the model to overfit the data. 

#### SOLUTION ENDS HERE

## Instructions for Question 3 (12 points)

- Company ABC is investing in Facebook, Pinterest, Twitter, and online display advertising to boost sales.
- To evaluate the effectiveness of their marketing budget, we are tasked with developing a straightforward Media Mix Model using linear regression methods.

<br />
The data for these questions is contained in the file, **`mmmdata.csv`**, which includes the following fields:

- **`Date`** : Date.
- **`Sales`** : Sales of Company ABC.
- **`Facebook`** : Marketing spending on Facebook.
- **`Pinterest`** : Marketing spending on Pinterest.
- **`Twitter`** : Marketing spending on Twitter.
- **`Online_Display`** : Marketing spending on Online_Display.


In [58]:
## Load dataset 
data_Q3 = safe_read_csv("../resource/asnlib/publicdata/mmmdata.csv", stringsAsFactors = FALSE)

## AdStock Terms
data_Q3$Facebook.adstock <- as.numeric(stats::filter(x= data_Q3$Facebook , filter =0.3 , method ="recursive"))
data_Q3$Pinterest.adstock <- as.numeric(stats::filter(x= data_Q3$Pinterest , filter =0.2 , method ="recursive"))
data_Q3$Twitter.adstock <- as.numeric(stats::filter(x= data_Q3$Twitter , filter =0.7 , method ="recursive"))
data_Q3$Online_Display.adstock <- as.numeric(stats::filter(x= data_Q3$Online_Display , filter =0.5 , method ="recursive"))

head(data_Q3)

[1] "../resource/asnlib/publicdata/mmmdata.csv  successfully read."
[1] "If  ../resource/asnlib/publicdata/mmmdata.csv  not found, please notify us in Piazza. Thank you."


Date    Sales  Facebook Pinterest Twitter Online_Display Facebook.adstock
1 5/23/32 570606 107019    69188    0       5468           107019.00       
2 5/24/32 660011 118857    69661    0       5373           150962.70       
3 5/25/32 416314 122417    13884    0       5524           167705.81       
4 5/26/32 443993 115249    41626    0       5380           165560.74       
5 5/27/32 757152  72823   119381    0       5229           122491.22       
6 5/28/32 641014  35275   109026    0       5189            72022.37       
  Pinterest.adstock Twitter.adstock Online_Display.adstock
1  69188.00         0                5468.00              
2  83498.60         0                8107.00              
3  30583.72         0                9577.50              
4  47742.74         0               10168.75              
5 128929.55         0               10313.38              
6 134811.91         0               10345.69

### A. Please develop the following  model. (4 points)

- **Model 1** : Regress `Sales` on Company ABC’s Facebook, Pinterest, Twitter, and Online Display spending.


In [60]:
# SOLUTION BEGINS HERE
sales_model <- lm(Sales ~ Facebook + Pinterest + Twitter + Online_Display, data = data_Q3)
summary(sales_model)
# SOLUTION ENDS HERE


Call:
lm(formula = Sales ~ Facebook + Pinterest + Twitter + Online_Display, 
    data = data_Q3)

Residuals:
    Min      1Q  Median      3Q     Max 
-254616 -110484   -5757  111865  301767 

Coefficients:
                 Estimate Std. Error t value Pr(>|t|)    
(Intercept)     6.006e+05  3.822e+05   1.571  0.11854    
Facebook        1.875e-02  2.151e-01   0.087  0.93069    
Pinterest      -8.678e-01  2.955e-01  -2.937  0.00392 ** 
Twitter         8.447e+00  1.698e+00   4.973 2.02e-06 ***
Online_Display  4.234e+00  7.135e+01   0.059  0.95276    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 139200 on 131 degrees of freedom
Multiple R-squared:  0.1867,	Adjusted R-squared:  0.1619 
F-statistic: 7.518 on 4 and 131 DF,  p-value: 1.75e-05


### B. Please develop the following  model. (4 points)

- **Model 2** : Regress `Sales` on the adstock terms defined in the instruction. **DO NOT** make any changes to the **`.adstock`** variables. **ONLY** include the **.adstock** variables listed in your analysis.


In [62]:
# SOLUTION BEGINS HERE
sales_adstock_model <- lm(Sales ~ Facebook.adstock + Pinterest.adstock + Twitter.adstock + Online_Display.adstock, data = data_Q3)
summary(sales_adstock_model)

# SOLUTION ENDS HERE


Call:
lm(formula = Sales ~ Facebook.adstock + Pinterest.adstock + Twitter.adstock + 
    Online_Display.adstock, data = data_Q3)

Residuals:
    Min      1Q  Median      3Q     Max 
-242004 -110880  -14999  107700  287927 

Coefficients:
                         Estimate Std. Error t value Pr(>|t|)    
(Intercept)             6.772e+05  2.329e+05   2.907  0.00428 ** 
Facebook.adstock        6.679e-02  2.019e-01   0.331  0.74127    
Pinterest.adstock      -6.626e-01  2.470e-01  -2.682  0.00826 ** 
Twitter.adstock         3.175e+00  5.756e-01   5.516 1.78e-07 ***
Online_Display.adstock -7.068e+00  2.184e+01  -0.324  0.74676    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 136400 on 131 degrees of freedom
Multiple R-squared:  0.219,	Adjusted R-squared:  0.1951 
F-statistic: 9.182 on 4 and 131 DF,  p-value: 1.43e-06


### C. Explain which model best explains the data and why? (4 points)

#### SOLUTION BEGINS HERE (Double click to edit)

The ad stock model models the data better because it has a better adjusted r squared value. In other words, the ad stock factors explain the variance in sales more than the original model. 

#### SOLUTION ENDS HERE